packages/jest-haste-map/src/index.js

```diff
@@ -0,0 +1,241 @@
+
+    return Promise.resolve()
+      .then(() => read.call(this))
+      .catch(() => this._createEmptyMap())
+      .then(data => this._crawl(data));
+  }
+
+  _buildHasteMap(data) {
+    const map = Object.create(null);
+    const setModule = module => {
+      if (!map[module.id]) {
+        map[module.id] = Object.create(null);
+      }
+      const moduleMap = map[module.id];
+      const platform = getPlatformExtension(module.path) || GENERIC_PLATFORM;
+      const existingModule = moduleMap[platform];
+      if (existingModule && existingModule.path !== module.path) {
+        console.warn(
+          `@providesModule naming collision:\n` +
+          `  Duplicate module name: ${module.id}\n` +
+          `  Paths: ${module.path} collides with ${existingModule.path}\n\n` +
+          `This warning is caused by a @providesModule declaration ` +
+          `with the same name accross two different files.`
+        );
+      }
+
+      const fileData = data.files[module.path];
+      fileData.id = module.id;
+      moduleMap[platform] = module;
+    };
+
+    const promises = [];
+    for (const filePath in data.files) {
+      if (!this._isNodeModulesDir(filePath)) {
+        const fileData = data.files[filePath];
+        const moduleData = data.map[fileData.id];
+        if (fileData.visited) {
+          if (!fileData.id) {
+            continue;
+          } else if (fileData.id && moduleData) {
+            map[fileData.id] = moduleData;
+            continue;
+          }
+        }
+
+        fileData.visited = true;
+        if (filePath.endsWith(PACKAGE_JSON)) {
+          promises.push(this._processHastePackage(filePath, setModule));
+        } else {
+          promises.push(this._processHasteModule(filePath, setModule));
+        }
+      }
+    }
+
+    return Promise.all(promises).then(() => {
+      data.map = map;
+      return data;
+    });
+  }
+
+  _parse(data) {
+    data = JSON.parse(data);
+    for (const key in data) {
+      Object.setPrototypeOf(data[key], null);
+    }
+    return data;
+  }
+
+  _crawl(data) {
+    const crawl =
+      (canUseWatchman && this._options.useWatchman) ? watchmanCrawl : nodeCrawl;
+
+    return crawl(
+      this._options.roots,
+      this._options.extensions,
+      this._options.ignorePattern,
+      data
+    );
+  }
+
+  _isNodeModulesDir(filePath) {
+    if (!filePath.includes(NODE_MODULES)) {
+      return false;
+    }
+
+    if (this._whitelist) {
+      const match = filePath.match(this._whitelist);
+      return !match || match[1].includes(NODE_MODULES);
+    }
+
+    return true;
+  }
+
+  _createEmptyMap() {
+    return {
+      clocks: Object.create(null),
+      files: Object.create(null),
+      map: Object.create(null),
+    };
+  }
+
+  _processHastePackage(filePath, setModule) {
+    return readFile(filePath, 'utf-8')
+      .then(data => {
+        data = JSON.parse(data);
+        if (data.name) {
```

Comment:
This is only really for haste-packages which is for package.json files that are not part of a node_modules folder. Jest does its own resolution, so we don't need to read all packages. The haste packages all have a name, so this should be fine.